### PR TITLE
Added "same" padding for conv*d

### PR DESCRIPTION
**Pull request by Chillee**
https://github.com/pytorch/pytorch/issues/3867

The implementation is taken from here (with some comments annotating the logic): https://github.com/mlperf/inference/blob/master/others/edge/object_detection/ssd_mobilenet/pytorch/utils.py#L40

In addition, I've stress tested it for all combinations of input image, kernel size, stride, and dilation from 1-6 (6^8 combinations). For all of them, the output dimensions follow the rules listed here for "same": https://www.tensorflow.org/api_docs/python/tf/nn/convolution

> `output_spatial_shape[i] = ceil(input_spatial_shape[i] / strides[i])`

~~I still need to add tests and test JIT compatability.~~ Then I can tackle the other ones.

### CODE REVIEW

File: torch/nn/modules/conv.py
Comment (fmassa) [inline torch/nn/modules/conv.py]: This means that users of `torch.nn.functional.conv2d` won't be able to use the `"same"` padding?

**Review by fmassa — CHANGES_REQUESTED** [review]
From my view, I think we should have 1:1 parity between the functional and the module interface.

This means that we should probably push the `"same"` handling down to the functional interface, but I'll let the others chime in as well.

**Comment by Chillee** [comment]
@fmassa I initially did it this way as it seems that Soumith implied that it should be separate from the functional interface: https://github.com/pytorch/pytorch/issues/3867#issuecomment-508322941

I continued doing it this way because it seemed like there already were differences between the functional and module interfaces. For example, the module interface supports an argument `padding_mode` that can be `circular` or `zeros`. I assumed that things were done this way as `circular` padding also requires a separate `F.pad` before the conv is called. 

Putting it in the functional interface also requires delving substantially deeper into the darker sides of Pytorch, and it wasn't clear to me where the appropriate location was, nor whether it would break something to have one op actually become 2 ops.

**Comment by apaszke** [comment]
Uhhh are we sure we want this? I remember @Yangqing _strongly_ arguing against adding this option as it's often quite surprising.

**Comment by Chillee** [comment]
@apaszke Do you remember any of the reasons he gave for strongly warning against this? 

@soumith mentioned that dynamically calculating padding is bad for a "variety of efficiency/serialization reasons", but these concerns don't seem relevant for research users. 

On the other hand, for research users this is an unnecessary pain point. I've had multiple people complain that this isn't in Pytorch. I'm not convinced that the performance/serialization issues are worth the cost here.

**Comment by soumith** [comment]
yes, we want this. YQ's concerns were around exported models not being static, and the ONNX export pass needs to error out if the padding info cannot be statically determined (@Chillee you might have to  add a test for the ONNX export).

At this point, the user need and benefits outweigh not wanting it, and our JIT / production export now can comfortably handle dynamic models (it wasn't true when we last thought about this)

**Comment by fmassa** [comment]
@Chillee I wasn't aware that `circular` padding was added only in the `nn.Module` interface.

I'm less sure we want to diverge the functionalities of both the module and functional interfaces, because some things are only possible to be done with the functional interface (for example, when we want to have the weights of a model to be the output of a network).

But I'll let @soumith and @apaszke chime in on their views

**Comment by soumith** [comment]
my main thinking around matching everything in the functional interface was around how much state we can carry to reuse previous calculations. We can carry state across `forward` calls in the Module, but not in the Functional interface.

I didn't explore the line of thought around how much state (such as pre-calculating padding values etc.) can be carried across iterations and how much benefits it brings.

**Comment by Chillee** [comment]
@fmassa One issue with putting it in the functional interface is that the semantics around auto-calculating "static padding" become fairly muddled. Currently, we can auto-calculate static padding and have it for free if certain rules are followed (stride=1, (filter-1)*dilation is divisible by 2). 

It's not clear how we would do this in the functional interface.

I think the right place to add this in the functional interface would be in the `F.pad` function or something along those lines.

**Comment by Chillee** [comment]
Added tests and fixed JIT compatibility. Removed the `WIP` tag - I think this is ready for review.

**Comment by fmassa** [comment]
@Chillee 
> One issue with putting it in the functional interface is that the semantics around auto-calculating "static padding" become fairly muddled. Currently, we can auto-calculate static padding and have it for free if certain rules are followed (stride=1, (filter-1)*dilation is divisible by 2).

I don't quite get this. The overhead of calculation of if the padding is static or not is negligible wrt the convolution time, so doing this in the functional codepath would not be a problem at all.

> It's not clear how we would do this in the functional interface.

Here is an example
```python
def _conv2d(...):
     # same as conv2d now

def conv2d(...):
    is_static, padding = _compute_if_padding_is_static(...)
    if is_static:
        input = pad(input, padding, etc)
        return _conv2d(input, ...)
    return _conv2d(input, ...)
```

> I think the right place to add this in the functional interface would be in the F.pad function or something along those lines.

I actually think we should instead add support for different paddings on left / right / top / bottom. This would be a generalization of what we currently have.
But this is a bit more involved to do.

**Comment by Chillee** [comment]
@fmassa I guess it wasn't clear to me what "static padding" actually meant. If that's still considered static padding, then I'm fine with doing it in the functional interface. Would you be able to provide a pointer to what I should be changing?

Also, as a side note, the documentation for `F.conv2d` isn't quite correct: https://pytorch.org/docs/stable/nn.html#id22

It says it supports `padding_mode` but it doesn't.

**Comment by Chillee** [comment]
So, two steps forward, one step back. I implemented the logic in the functional interface, but now I've run into 2 problems:
1. I needed to comment out [this code](https://github.com/pytorch/pytorch/pull/22484/files#diff-54816789e967b71e1be8048e53f8ccbdL305). Otherwise, it'd complain that `invalid default string: "same"`.
2. Things no longer work with the JIT, due to the fact that padding can be either a string or a tuple. There's 3 different ways of handling this, that I see:
a. Do what's being done [here](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/rnn.py#L523) to overload the function.
b. Add another atttribute to the conv module that controls whether or not we use `padding="same"`. In this way, we can simply always check that attribute to see whether we pass in `"same"` or `self.padding`.
c. Port `circular` over to `F.pad` as well.

I like `c` the most, since it also resolves another issue.

**Comment by Chillee** [comment]
So, two steps forward, one step back. I implemented the logic in the functional interface, but now I've run into 2 problems:
1. I needed to comment out [this code](https://github.com/pytorch/pytorch/pull/22484/files#diff-54816789e967b71e1be8048e53f8ccbdL305). Otherwise, it'd complain that `invalid default string: "same"`.
2. Things no longer work with the JIT, due to the fact that padding can be either a string or a tuple. There's 3 different ways of handling this that I see:
a. Do what's being done [here](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/rnn.py#L523) to overload the function.
b. Add another atttribute to the conv module that controls whether or not we use `padding="same"`. In this way, we can simply always check that attribute to see whether we pass in `"same"` or `self.padding`.
c. Port `circular` over to `F.pad` as well.

I like `c` the most, since it also resolves another issue.

What do you guys think?

@soumith @fmassa

**Comment by fmassa** [comment]
@Chilee what about option `d`: add support for padding differently in the left / right side of a convolution in C++ (via padding of the difference, e.g.,), and then on the python side you just need to compute the amount of padding needed if you pass `padding=same`?
This way, the interface on the C++ side is almost unchanged, it stays as tuple as before, and we just need to make add some extra logic in the python side?

Thoughts?

**Comment by soumith** [comment]
Francisco's option `d` seems clean / cleaner.

**Comment by Chillee** [comment]
@fmassa @soumith I do agree that putting the logic in `F.conv2d` is cleaner, but it doesn't solve the original issue posed - how do I solve the issue that Torchscript is statically typed, and can only take in parameters of one type?

I think this goes back to the original 3 options.

**Comment by Chillee** [comment]
The only way I could find to do this was to use the `weak_script` annotations and overload `F.conv2d`. However, I'm not sure if this is intended (I think my team has implied that it's not...)

**Comment by soumith** [comment]
@Chillee I think torchscript has to handle multiple input types, or keep around weakscript. I cant see an alternative, and we shouldn't design our APIs with torchscript limitations in mind.

File: torch/nn/modules/conv.py
Comment (eellison) [inline torch/nn/modules/conv.py]: `isinstance` metacompilation will only work in an if statement currently, not ternary expression

**Comment by spandantiwari** [comment]
> yes, we want this. YQ's concerns were around exported models not being static, and the ONNX export pass needs to error out if the padding info cannot be statically determined (@Chillee you might have to add a test for the ONNX export).
> 
> At this point, the user need and benefits outweigh not wanting it, and our JIT / production export now can comfortably handle dynamic models (it wasn't true when we last thought about this)

I agree with @soumith 's comment in that ONNX Conv op will support similar semantics as `same` being here, with the `auto_pad` option `SAME_UPPER`. This option was previously on path to deprecation, but due to clear use cases has since been retained in the design. Given that I feel that this can be handled for ONNX export.

**Comment by eellison** [comment]
@Chillee do you want someone else to take on finishing this ? It would be nice to get this in for 1.3

**Comment by Chillee** [comment]
@eellison I've been busy (and will be busy) until the 23rd, but I should be able to finish it up in the week after that (which should be at least a week before 1.3?).

**Comment by eellison** [comment]
I think the cut is the 24th actually... although potentially we could cherry-pick this in

**Comment by Chillee** [comment]
Ah alright I'll take a crack tonight at fixing it then.

**Comment by eellison** [comment]
@ZolotukhinM @jamesr66a @raghuramank100 this PR breaks convolution tests. Do you think it's doable to land this PR before the release and have quant parity with it ?

**Comment by Chillee** [comment]
To clarify a bit more, the tests that break are the ones that look like (the last line specifically).
```
        FileCheck().check('ClassType<Observer> = prim::GetAttr[name="observer_for_') \
                   .check_next('prim::CallMethod[name="forward"](%observer_for_') \
                   .check('ClassType<Observer> = prim::GetAttr[name="observer_for_') \
                   .check_next('prim::CallMethod[name="forward"](%observer_for_') \
                   .check('Tensor = aten::conv2d') \
```

The tests pass as long as I comment that out - it's not clear to me if that's the right thing to do.

**Comment by eellison** [comment]
@Chillee quantization has a mirror'd `conv` implementation so to maintain parity they have to do more than fix the tests. Since the branch cut is today/tomorrow quant people (cc @jamesr66a) think it would be better to wait until after the cut to land rather than rush in a quant fix. 

I forgot that quantization needed to maintain parity when I commented above about getting it into the release, that's my bad.

**Comment by Yangqing** [comment]
Not that it matters, but I have a python notebook detailing the design choices:

https://gist.github.com/Yangqing/47772de7eb3d5dbbff50ffb0d7a98964

As an architect, I still believe that the SAME padding is not something that we should introduce back into the common usage patterns. This being said, I was the person that included SAME paddings in ONNX for compatibility reasons, so I can understand why one might want it.

**Comment by ZolotukhinM** [comment]
I somehow missed the notification 16 days ago, sorry.

> To clarify a bit more, the tests that break are the ones that look like (the last line specifically).
> 
> ```
>         FileCheck().check('ClassType<Observer> = prim::GetAttr[name="observer_for_') \
>                    .check_next('prim::CallMethod[name="forward"](%observer_for_') \
>                    .check('ClassType<Observer> = prim::GetAttr[name="observer_for_') \
>                    .check_next('prim::CallMethod[name="forward"](%observer_for_') \
>                    .check('Tensor = aten::conv2d') \
> ```
> 
> The tests pass as long as I comment that out - it's not clear to me if that's the right thing to do.

Could you please paste what graph do we actually have there? To print it, you could add
```
print(str(m._c._get_module("conv")._get_method('conv2d_forward').graph))
```
before the `FileCheck` invocation.

I'm pretty sure this test can be adjusted to the changes you're doing - it basically just checks that after the transformation we have instrumentation calls to "observers" before the `conv` call that we're instrumenting. It essentially looks for specific strings and my guess is that `conv` now looks slightly differently.

**Comment by eellison** [comment]
@Chillee , @ZolotukhinM @jamesr66a what can we do to land this

**Comment by Chillee** [comment]
Sorry I've been busy recently - will get back to this within a week or two.

**Comment by Chillee** [comment]
Somebody else pushed it across the finish line when I couldn't :) https://github.com/pytorch/pytorch/pull/45667

**Pull request closed without merging**
